### PR TITLE
Add SequencesExt!Replace(First|All)SubSeq operators

### PR DESCRIPTION
**Pull request by lemmy**
Question: Do we want `ReplaceFirstSubSeq(<<1>>, <<>>, <<>>)` to equal `<<>>` or `<<1>>`?  The latter would be in line with e.g. Java's `String#replaceFirst`, and, thus, what users expect.

Note: Not using `DOMAIN` in the definition of the ops to [support strings](https://github.com/tlaplus/tlaplus/issues/512).  Module override will be added later.

(Inspired by a private patch shared by @afonsonf with me)

Partially related: https://github.com/tlaplus/CommunityModules/issues/40

### CODE REVIEW

**Review by afonsonf — APPROVED** [review]
LGTM

File: modules/SequencesExt.tla [hunk 1]
```diff
@@ -124,6 +124,37 @@ Last(s) ==
...
+  \E i,j \in 1..Len(t):
+    /\ i <= j \* Optimization to not eval SubSeq with j < i, which is defined as <<>>.
+    /\ SubSeq(t, i, j) = s
+
+ReplaceFirstSubSeq(r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with its subsequence  s  replaced by the sequence  r  *)
+  (**************************************************************************)
+  IF IsSubSequence(s, t) THEN \* Prevent CHOOSE x \in S: P not holding for any element.
+    LET idx == CHOOSE i \in 1..Len(t): \E j \in i..Len(t): SubSeq(t, i, j) = s
```
Comment (muenchnerkindl) [inline modules/SequencesExt.tla:140]: The name of the operator suggests that the first occurrence of s in t will be replaced by r, but the definition doesn't enforce this. One could either rename the operator or change the definition to

LET idx == CHOOSE i \in 1 .. Len(t) : 
                        /\ SubSeq(t, i, Len(s)) = s
                        /\ \A j \in 1 .. i-1 : SubSeq(t, j, Len(s)) # s

I presume that the way TLC implements CHOOSE ensures that the first match will be chosen, though.

File: modules/SequencesExt.tla [hunk 2]
```diff
@@ -124,6 +124,37 @@ Last(s) ==
...
+        suffix == r \o SubSeq(t, idx + Len(s), Len(t))
+    IN IF idx > 1
```
Comment (muenchnerkindl) [inline modules/SequencesExt.tla:142]: Is this case distinction really necessary? Don't we have SubSeq(t, 1, 0) = << >> ?

File: modules/SequencesExt.tla [hunk 3]
```diff
@@ -124,6 +124,37 @@ Last(s) ==
...
+       ELSE suffix
+  ELSE t
+
+RECURSIVE ReplaceAllSubSeqs(_,_,_)
+ReplaceAllSubSeqs(r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with all subsequences  s  replaced by the sequence  r *)
+  (**************************************************************************)
+  IF IsSubSequence(s, t)
+  THEN ReplaceAllSubSeqs(r, s, ReplaceFirstSubSeq(r, s, t))
```
Comment (muenchnerkindl) [inline modules/SequencesExt.tla:153]: I expect ReplaceAllSubSeqs(r, s, s) to loop, just like ReplaceAllSubSeqs(r, <<>>, t), right? Perhaps add a warning to the comment?

**Review by muenchnerkindl — COMMENTED** [review]
I would expect ReplaceFirstSubSequence(<<1>>, <<>>, <<>>) to return <<1>>.

File: modules/SequencesExt.tla [hunk 4]
```diff
@@ -124,6 +124,53 @@ Last(s) ==
...
+ReplaceFirstSubSeq(r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with its subsequence  s  replaced by the sequence  r  *)
+  (**************************************************************************)
+  IF s \in SubSeqs(t)
+  THEN ReplaceSubSeqAt(IndexFirstSubSeq(s, t), r, s, t)
+  ELSE t
+
+RECURSIVE ReplaceAllSubSeqs(_,_,_)
+ReplaceAllSubSeqs(r, s, t) ==
```
Comment (muenchnerkindl) [inline modules/SequencesExt.tla:161]: Looks good to me, but I believe that

ReplaceAllSubSeqs(<<1,1>>, <<1>>, <<1,1>>)

loops, this is probably what is expected? Perhaps clarify in a comment. 
Alternatively, one would have to keep track of the starting index for replacement in the recursive call in order to obtain

ReplaceAllSubSeqs(<<1,1>>, <<1>>, <<1,1>>) = <<1,1,1,1>>

**Comment by lemmy** [comment]
@muenchnerkindl Can you please review the (non-recursive) `ReplaceAllSubSeqs` in https://github.com/tlaplus/CommunityModules/pull/52/commits/0f53203e8a52d46097979140e1885c30684f1ec4 ?

**Review by muenchnerkindl — APPROVED** [review]
Whew: ReplaceAllSubSeqs looks complicated but I cannot find a fault with it and I couldn't come up with a better suggestion either.

**Comment by lemmy** [comment]
Thanks for the reviews!

By the way, I learned that Java and Dotnet handle the empty sequence differently in their replace* methods.

**Pull request merged**